src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,223 @@
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const code_function_callt &function_call =
+          to_code_function_call(instr.code);
+        DATA_CHECK(
+          vm,
+          function_call.function().id() == ID_symbol,
+          "no calls via function pointer should be present");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_returns_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    const goto_functiont &goto_function = fun.second;
+    DATA_CHECK(
+      vm,
+      goto_function.type.return_type().id() == ID_empty,
+      "functions must have empty return type");
+
+    for(const auto &instr : goto_function.body.instructions)
+    {
+      DATA_CHECK(
+        vm, !instr.is_return(), "no return instructions should be present");
+
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        DATA_CHECK(
+          vm,
+          function_call.lhs().is_nil(),
+          "function call return should be nil");
+
+        const auto &callee = to_code_type(function_call.function().type());
+        DATA_CHECK(
+          vm,
+          callee.return_type().id() == ID_empty,
+          "called function must have empty return type");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_called_functions()
+{
+  class test_for_function_addresst : public const_expr_visitort
+  {
+  public:
+    test_for_function_addresst(
+      const function_mapt &function_map,
+      const validation_modet &vm)
+      : function_map{function_map}, vm{vm}
+    {
+    }
+
+    void operator()(const exprt &expr) override
+    {
+      if(expr.id() == ID_address_of)
+      {
+        const auto &pointee = to_address_of_expr(expr).object();
+        if(pointee.id() == ID_symbol)
+        {
+          const auto &identifier = to_symbol_expr(pointee).get_identifier();
+          DATA_CHECK(
+            vm,
+            function_map.find(identifier) != function_map.end(),
+            "every function whose address is taken must be in the "
+            "function map");
+        }
+      }
+    }
+
+  private:
+    const function_mapt &function_map;
+    const validation_modet &vm;
+  };
+  test_for_function_addresst test_for_function_address(function_map, vm);
+
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      // check functions that are called
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        const irep_idt &identifier =
+          to_symbol_expr(function_call.function()).get_identifier();
+
+        DATA_CHECK(
+          vm,
+          function_map.find(identifier) != function_map.end(),
+          "every function call callee must be in the function map");
+      }
+
+      // check functions of which the address is taken
+      const exprt &src{instr.code};
```

Comment:
:pick: Use assignment here, something `const auto &src = static_cast<const exprt&>(instr.code);`. As is this imho looks a bit too much like a copy constructor.

Reply:
Done.